server/statshouse/worker-stats-buffer.cpp

```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+void WorkerStatsBuffer::make_metric(std::vector<StatsHouseMetric> &metrics, const char *name, QueryStatKey stat_key, const std::vector<tag> &tags) {
+  auto &stats_buffer = query_stats[static_cast<size_t>(stat_key)];
+  if (!stats_buffer.empty()) {
+    metrics.push_back(make_statshouse_value_metrics(name, stats_buffer.get_data_and_reset_buffer(), tags));
+  }
+}
+
+void WorkerStatsBuffer::flush() {
+  dl::CriticalSectionGuard critical_section;
+  std::vector<StatsHouseMetric> metrics;
+  const char* hostname = kdb_gethostname();
+
+  for (size_t i = 0; i < static_cast<size_t>(WorkerType::types_count); ++i) {
+    std::vector<tag> tags;
+    tags.emplace_back("worker_type", i == static_cast<size_t>(WorkerType::general_worker) ? "general" : "job");
+    if (hostname != nullptr) {
+      tags.emplace_back("host", kdb_gethostname());
+    }
+
+    make_generic_metric(metrics, "kphp_requests_outgoing_queries", GenericQueryStatKey::outgoing_queries, i, tags);
+    make_generic_metric(metrics, "kphp_requests_outgoing_long_queries", GenericQueryStatKey::outgoing_long_queries, i, tags);
+    make_generic_metric(metrics, "kphp_requests_net_time", GenericQueryStatKey::script_time, i, tags);
+    make_generic_metric(metrics, "kphp_requests_script_time", GenericQueryStatKey::net_time, i, tags);
+  }
+
+  std::vector<tag> tags;
+  if (hostname != nullptr) {
+    tags.emplace_back("host", hostname);
+  }
+  make_metric(metrics, "kphp_memory_script_usage", QueryStatKey::general_memory_used, tags);
+  make_metric(metrics, "kphp_memory_script_real_usage", QueryStatKey::general_real_memory_used, tags);
+  make_metric(metrics, "kphp_memory_script_total_allocated_by_curl", QueryStatKey::general_total_allocated_by_curl, tags);
+
+  make_metric(metrics, "kphp_jobs_queue_time", QueryStatKey::job_wait_time, tags);
+  make_metric(metrics, "kphp_memory_job_request_usage", QueryStatKey::job_request_memory_usage, tags);
+  make_metric(metrics, "kphp_memory_job_request_real_usage", QueryStatKey::job_request_real_memory_usage, tags);
+  make_metric(metrics, "kphp_memory_job_response_usage", QueryStatKey::job_response_memory_usage, tags);
+  make_metric(metrics, "kphp_memory_job_response_real_usage", QueryStatKey::job_response_real_memory_usage, tags);
+
+  make_metric(metrics, "kphp_memory_job_common_request_usage", QueryStatKey::job_common_request_memory_usage, tags);
+  make_metric(metrics, "kphp_memory_job_common_request_real_usage", QueryStatKey::job_common_request_real_memory_usage, tags);
+
+  int offset = vk::tl::store_to_buffer(buffer, buffer_size, TL_STATSHOUSE_ADD_METRICS_BATCH);
+  offset = offset + vk::tl::store_to_buffer(buffer + offset, buffer_size - offset, static_cast<int>(vk::tl::statshouse::add_metrics_batch_fields_mask::ALL));
+  int len = vk::tl::store_to_buffer(buffer + offset, buffer_size - offset, metrics);
+  vk::singleton<StatsHouseClient>::get().send_metrics(buffer, offset + len);
+}
+
+void WorkerStatsBuffer::flush_if_needed() {
```

Comment:
Why do we need to call this function in `cron`, if we always check time at `add_query_stat`?